Привет, меня зовут Гребенюк Артём. Сегодня я буду делать ревью твоего проекта!\
В Практикуме принято общаться на "ты", однако если это неудобно - дай знать, и можем перейти на более официальное общение.

Ниже ты найдешь мои комментарии - пожалуйста, не перемещай, не изменяй и не удаляй их. Я буду использовать цветовую разметку: 

<div class="alert alert-block alert-success">
<b>✔️ Success:</b> Если все сделано хорошо. </div>

<div class="alert alert-block alert-warning">
<b>⚠️ Remarks:</b> Некритичное замечание; если таких замечаний меньше трех — проект может быть принят без их отработки.</div>

<div class="alert alert-block alert-danger">
<b>❌ Needs fixing:</b> Грубая ошибка, необходимо исправить, иначе проект не будет принят.
</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
Например так:

<div class="alert alert-block alert-info">
<b>💾 Fixed:</b> 
Hello World! ✔
</div>

Или можешь использовать эту галочку ✔ в моих комментариях, если писать особо нечего.

<div class="alert alert-block alert-warning">
<b>Общее впечатление:</b> 
    
- Спасибо за работу :)
- У меня есть пара ремарок, которые помогут тебе сделать проект еще лучше.
- Пробегись по моим комментариям, исправь недочёты и всё будет отлично :)
    
</div>



<div class="alert alert-block alert-success">
<b>Привет!</b> Отличная работа над ремарками! Проект принят, поздравляю, желаю удачи в дальнейшем обучении :)

# Исследование надежности заемщиков.

<div class="alert alert-block alert-warning">
<b>Remarks:</b>

Любому проекту нужно введение. Также стоит задуматься, как еще его можно дополнить, чтобы сделать более ценным для читателя. В студенческом ноушене (если нет доступа, спроси куратора) есть статья о том, как мы оформляем работы.
    
Также советую тебе взглянуть на <a href="https://www.kaggle.com/notebooks/">работы</a> профессиональных аналитиков, чтобы оценить, как другие люди оформляют и проводят исследования. </div>

<div class="alert alert-block alert-warning">
<b>Задача:</b>

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.
 </div>

### Шаг 1. Обзор данных

In [1]:
import pandas as pd
banki = pd.read_csv('/datasets/data.csv')

<div class="alert alert-block alert-success">
    
- тут следовало импорт поместить в отдельную ячейку. Было бы импортов несколько - все их в первую ячейку, а уже потом открытие датасета и работа с ним
- в следующей ячейке было бы удобнее применить что-нибудь из `head()/tail()/smaple()` для просмотра части датасета

In [2]:
banki

children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

<div class="alert alert-block alert-warning">
<b>⚠️ Remarks:</b>
    
- стоит отделять импорты от открытия датасета. Это делается для того, чтобы иметь возможность в любой момент дописать новые библиотеки, настройки или ключевые константы, перезапустить ячейку с импортами и продолжить работу, не перетирая датасет.
- если в аутпут попадает результат только одной функции, то ни `print()`, ни `display()` не нужны - юпитер всё отрисует сам

### Шаг 2.1 Заполнение пропусков

Беру основные сведения о таблице. Задача: надо выяснить, влияет ли семейное положение и количество детей клиента на факт возврата кредита в срок? В выводе видно что занчения в столбцах days_employed и total_income отличаются от большинства, значит в них есть пропущеные значения. 

In [3]:
banki.info() 

banki_days_employed = banki['children'].count() - banki['days_employed'].count()
print("Количество пустых значений в кол-ве рабочих дней равно", banki_days_employed)
banki_total_income = banki['children'].count() - banki['total_income'].count()
print("Количество пустых значений в общей прибыли равно", banki_total_income)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
Количество пустых значений в кол-ве рабочих дней равно 2174
Количество пустых значений в общей прибыли равно 2174


In [4]:
#посмотрим что в долях
days_employed_skip = banki_days_employed / len(banki)
total_income_skip = banki_total_income/ len(banki)

print ('Доля пустых значений в кол-ве рабочих дней равна:', days_employed_skip) #нашла долю пустых значений столбца days_employed
print ('Доля пустых значений в общей прибыли равна:', total_income_skip) #нашла долю пустых значений столбца total_income

Доля пустых значений в кол-ве рабочих дней равна: 0.10099883855981417
Доля пустых значений в общей прибыли равна: 0.10099883855981417


Странно что это кол-во совпадает. И так же странно что у обоих столбцов тип данных - float64. М.б. в этом проблема? Если бы это были все безработные люди , у них нет работы и поэтому пусто в колонке total_income - то такая гипотеза 
верна едвали, т.к. у кого то доложен быть доход помимо основной работы. Скорее всего речь об ошибке в БД, поэтому правильнее всего было бы заполнить пропуски медианным значением:

<div class="alert alert-block alert-warning">
<b>⚠️ Remarks:</b> 
    
- старайся пояснения к исследованию всегда писать в маркдаун-ячейках. Через # Обычно пишутся технические комментарии к нетривиальным участкам кода
- будет лучше разбить код на большее число блоков, перемежающихся маркдаун-ячейками с комментариями

In [5]:
days_employed_median = banki[banki['days_employed'] < 0] 
print(days_employed_median['days_employed'].median())
total_income_median = banki[banki['total_income'] < 0]
print(total_income_median['total_income'].median())

-1630.0193809778218
nan


Вижу отрицательное медианное значение в столбце days_employed, с этим как раз разбираюсь в следующем пункте проекта.

<div class="alert alert-block alert-warning">
<b>⚠️ Remarks:</b> 
    
давай попробуем заполнить пропуски ориентируясь на какие-нибдуь характеристики заемщиков -- пол/возраст/тип_занятости

In [6]:
banki.columns #на всякий пожарный посмотрела пробелы

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

### Шаг 2.2 Проверка данных на аномалии и исправления.

Ищу аномалии. Итак, проверим на суммарное количество пропусков:

In [7]:
banki.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

По каким-то причинам много пропусков, опять же как писала выше, кол-во рабочих дней - возможно нет работы у человека. 
Однако в столбце доход - пропуски почти все, это не может быть случайностью, здесь техническая ошибка, инфасотка.
Может проблема с записью данных или типа того.

Тогда сначала заменю пропущеные значения. Потом смотрю опять что нет пропусков в таблице

In [8]:
banki['days_employed'] = banki['days_employed'].fillna(value = banki['days_employed'].mean())
banki['total_income'] = banki['total_income'].fillna(value = banki['total_income'].median())

In [9]:
banki.isnull().sum() # О, нормас теперь

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

In [10]:
for row in banki: 
  print(banki[row].value_counts())

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64
 63046.497661     2174
-327.685916          1
-2186.168272         1
-1580.622577         1
 361900.083341       1
                  ... 
-7120.517564         1
-2146.884040         1
-881.454684          1
-794.666350          1
-3382.113891         1
Name: days_employed, Length: 19352, dtype: int64
35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74   

Выводы:
"children" есть значения "20" и -1, поменять на 2 и 1 соответственно.
"days_employed" есть отрицательные значения. Надо изменить чтоб было только Int.Пропущено 19352
"education" названия задублированы, надо привести к единому регистру.
"gender"  пропущеные  
"total_income". Пропущенные значения (19352) - пока хз как их вставить
"education"  дубликаты, привести к одному регистру.

In [11]:
# children
banki['children'] = banki['children'].replace(20, 2)
banki['children'] = banki['children'].replace(-1, 1)
banki.groupby('children')['income_type'].count()

children
0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: income_type, dtype: int64

<div class="alert alert-block alert-success">
<b>✔️ Success:</b> замечательно!

In [12]:
# days_employed

banki['days_employed'] = banki['days_employed'].astype('int') #перевожу в инт
banki['days_employed'] = banki['days_employed'].abs() #убираю отрицательные значения
banki.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  int64  
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


In [13]:
banki['days_employed'].head(20)

0       8437
1       4024
2       5623
3       4124
4     340266
5        926
6       2879
7        152
8       6929
9       2188
10      4171
11       792
12     63046
13      1846
14      1844
15       972
16      1719
17      2369
18    400281
19     10038
Name: days_employed, dtype: int64

In [14]:
# Gender
# количество пропущенных значений:
banki[banki['gender'] == 'XNA'].count()
# исключу  
banki = banki[banki['gender'] != 'XNA']
banki.groupby('gender')['income_type'].count()

gender
F    14236
M     7288
Name: income_type, dtype: int64

In [43]:
# total_income
# пропущенные значения - оставлю пока хз как делать
pivot_table_total_income = banki.pivot_table(index=['dob_years'],
                                    columns = 'gender',
                                    values = 'total_income',
                                    aggfunc='median')

def fillna_in_total_income(banki):
    gender = banki['gender']
    age = banki['dob_years']
    if pd.isnull(banki['total_income']):
        return pivot_table_total_income.loc[age,gender]
    else:
        return banki['total_income']
    
    
banki['total_income'] = banki.apply(fillna_in_total_income, axis = 1)
banki.groupby('total_income')['income_type'].count()
banki.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   children               21453 non-null  int64 
 1   days_employed          21453 non-null  int64 
 2   dob_years              21453 non-null  int64 
 3   education              21453 non-null  object
 4   education_id           21453 non-null  int64 
 5   family_status          21453 non-null  object
 6   family_status_id       21453 non-null  int64 
 7   gender                 21453 non-null  object
 8   income_type            21453 non-null  object
 9   debt                   21453 non-null  int64 
 10  total_income           21453 non-null  int64 
 11  purpose                21453 non-null  object
 12  total_income_category  21452 non-null  object
 13  purpose_category       21453 non-null  object
 14  type_of_family         21453 non-null  object
dtypes: int64(7), object

In [44]:

# education 
# к одному регистру
banki['education'] = banki['education'].str.lower()
# теперь 5 уникальных значений
banki['education'].unique()
banki.groupby('education')['income_type'].count()

education
высшее                  5250
начальное                282
неоконченное высшее      743
среднее                15172
ученая степень             6
Name: income_type, dtype: int64

<div class="alert alert-block alert-success">
<b>✔️ Success:</b> правильно, избавились от неявных дубликатов

### Шаг 2.3. Изменение типов данных.

Для изменения типа данных используем метод astype() так как нам нужно изменить float на int64

In [46]:
# Изменим значения столбца total_income в числовой тип int64 (целое число) 
banki['total_income'] = banki['total_income'].astype('int')

In [48]:
# Изучим 
banki.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   children               21453 non-null  int64 
 1   days_employed          21453 non-null  int64 
 2   dob_years              21453 non-null  int64 
 3   education              21453 non-null  object
 4   education_id           21453 non-null  int64 
 5   family_status          21453 non-null  object
 6   family_status_id       21453 non-null  int64 
 7   gender                 21453 non-null  object
 8   income_type            21453 non-null  object
 9   debt                   21453 non-null  int64 
 10  total_income           21453 non-null  int64 
 11  purpose                21453 non-null  object
 12  total_income_category  21452 non-null  object
 13  purpose_category       21453 non-null  object
 14  type_of_family         21453 non-null  object
dtypes: int64(7), object

<div class="alert alert-block alert-warning">
<b>⚠️ Remarks:</b>
    
Потенциал для эффективного изменения типов тут очень широк.
    
Давай я дам тебе две классные статьи, а ты попробуешь изменить типы данных на еще более подходящие и замеришь, сколько памяти в % ты сэкономишь:
    
https://pyprog.pro/data_types.html
    
https://habr.com/ru/company/ruvds/blog/442516/

In [49]:
banki.info()
# <получение общей информации о данных таблицы df>

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   children               21453 non-null  int64 
 1   days_employed          21453 non-null  int64 
 2   dob_years              21453 non-null  int64 
 3   education              21453 non-null  object
 4   education_id           21453 non-null  int64 
 5   family_status          21453 non-null  object
 6   family_status_id       21453 non-null  int64 
 7   gender                 21453 non-null  object
 8   income_type            21453 non-null  object
 9   debt                   21453 non-null  int64 
 10  total_income           21453 non-null  int64 
 11  purpose                21453 non-null  object
 12  total_income_category  21452 non-null  object
 13  purpose_category       21453 non-null  object
 14  type_of_family         21453 non-null  object
dtypes: int64(7), object

In [50]:
banki['dob_years'] = banki['dob_years'].astype('int')
banki.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   children               21453 non-null  int64 
 1   days_employed          21453 non-null  int64 
 2   dob_years              21453 non-null  int64 
 3   education              21453 non-null  object
 4   education_id           21453 non-null  int64 
 5   family_status          21453 non-null  object
 6   family_status_id       21453 non-null  int64 
 7   gender                 21453 non-null  object
 8   income_type            21453 non-null  object
 9   debt                   21453 non-null  int64 
 10  total_income           21453 non-null  int64 
 11  purpose                21453 non-null  object
 12  total_income_category  21452 non-null  object
 13  purpose_category       21453 non-null  object
 14  type_of_family         21453 non-null  object
dtypes: int64(7), object

In [51]:
#Вроде получше стало, да?

<div class="alert alert-block alert-success">
да, но есть еще потенциал - можно уменьшить разрядность числовых столбцов и прмиенить тип category для строк
    

### Шаг 2.4. Удаление дубликатов.

In [20]:
banki.duplicated().sum() # дубликатов полно конечно тут у вас ))

71

<div class="alert alert-block alert-warning">
<b>⚠️ Remarks:</b>
    
это не ошибка, но важно понимать, что  у нас нет никакого уникального идентификатора строки в датасете => мы не можем с уверенностью сказать, дубликаты этоиил совпадения

In [21]:
banki = banki.drop_duplicates().reset_index(drop=True) # удаляю

In [22]:
banki.duplicated().sum() # жить можно

0

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [23]:
# иду по заданию, создаю датафреймы
education = banki[['education','education_id']]
education = education.sort_values('education').groupby('education').mean().reset_index()
education = education.drop('education', 1) #зачем то удаляем столбец, смысл этого мне пока непонятен, но делаю по заданию конечно, а то вы потом скажете почему не сделали
education

education_id
0             0
1             3
2             2
3             1
4             4

In [24]:
family = banki[['family_status','family_status_id']]
family = family.sort_values('family_status').groupby('family_status').mean().reset_index()
family = family.drop('family_status', 1)
family

family_status_id
0                 4
1                 3
2                 2
3                 1
4                 0

In [25]:
#Категоризация дохода

In [26]:
#группирую данные о доходе
def income(row):
    
    income = row['total_income']

    if income <= 30000:
        return 'E'
    if income <= 50000 and income > 30001:
        return 'D'
    if income <= 200000 and income > 50001:
        return 'C'
    if income <= 1000000 and income > 200001:
        return 'B'
    if income > 1000001:
        return 'A'
    
banki['total_income_category'] = banki.apply(income, axis=1)
banki

children  days_employed  dob_years education  education_id  \
0             1           8437         42    высшее             0   
1             1           4024         36   среднее             1   
2             0           5623         33   среднее             1   
3             3           4124         32   среднее             1   
4             0         340266         53   среднее             1   
...         ...            ...        ...       ...           ...   
21448         1           4529         43   среднее             1   
21449         0         343937         67   среднее             1   
21450         1           2113         38   среднее             1   
21451         3           3112         38   среднее             1   
21452         2           1984         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21448  гражданский брак                 1      F   компаньон     0   
21449   женат / замужем                 0      F   пенсионер     0   
21450  гражданский брак                 1      M   сотрудник     1   
21451   женат / замужем                 0      M   сотрудник     1   
21452   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  
0            253875                 покупка жилья                     B  
1            112080       приобретение автомобиля                     C  
2            145885                 покупка жилья                     C  
3            267628    дополнительное образование                     B  
4            158616               сыграть свадьбу                     C  
...             ...                           ...                   ...  
21448        224791             операции с жильем                     B  
21449        155999          сделка с автомобилем                     C  
21450         89672                  недвижимость                     C  
21451        244093  на покупку своего автомобиля                     B  
21452         82047         на покупку автомобиля                     C  

[21453 rows x 13 columns]

In [27]:
print(banki['purpose'].unique()) #слишко много вариаций

['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']


In [28]:
#создам столбец purpose_category
def purpose(row):
    
    purpose = row['purpose']

    if 'автомоби' in purpose:
        return "операции с автомобилем"
    if "образова" in purpose:
        return "получение образование"
    if "свадьб" in purpose:
        return "проведение свадьбы"
    if "недвижимост" or "жиль" in purpose:
        return "операции с недвижимостью"

banki['purpose_category'] = banki.apply(purpose, axis=1)
#banki.head(30)

In [29]:
banki.groupby('purpose_category')['purpose'].count()# недвижка чаще всего, ясно и так

purpose_category
операции с автомобилем       4306
операции с недвижимостью    10810
получение образование        4013
проведение свадьбы           2324
Name: purpose, dtype: int64

<div class="alert alert-block alert-success">
<b>✔️ Success:</b> отличная работа по разделу!

### Шаг 2.7. Категоризация целей кредита.

In [30]:
#разберемся с детьми. Делим всех кто сколько нарожал:
def count_of_child(row):
    if row == 0:
        return 'семьи без детей'
    if row <= 2 and row > 1:
        return 'семьи c 1 ребенком'
    if row <= 3 and row > 2:
        return 'семьи c 2 детьми'
    if row <= 4 and row > 3:
        return 'семьи c 3 детьми'
    else:
        return 'многодетные'

<div class="alert alert-block alert-success">
<b>✔️ Success:</b> хорошая категоризация! учитывая, что заемщиков с >2 детьми в датасете очень мало, я бы и семьи с 3 детьми в категорию "многодетные" включил

In [31]:
banki['type_of_family'] = banki['children'].apply(count_of_child)

In [32]:
banki_of_family = banki.groupby(['type_of_family','debt']).agg({'debt':["count"]})
banki_of_family

debt
                         count
type_of_family     debt       
многодетные        0      4419
                   1       445
семьи c 1 ребенком 0      1926
                   1       202
семьи c 2 детьми   0       303
                   1        27
семьи c 3 детьми   0        37
                   1         4
семьи без детей    0     13027
                   1      1063

In [33]:
banki_deti = banki.pivot_table(index=['debt'], columns='children', values = 'family_status_id', aggfunc='count')
b1 = banki_deti[0][1] / banki_deti[0][0]
b2 = banki_deti[1][1] / banki_deti[1][0]
b3 = banki_deti[2][1] / banki_deti[2][0]
b4 = banki_deti[3][1] / banki_deti[3][0]
b5 = banki_deti[4][1] / banki_deti[4][0]

print("{0:.2f}% Нет детей".format(b1*100))
print("{0:.2f}% 1 Ребенок".format(b2*100))
print("{0:.2f}% 2 Ребенка".format(b3*100))
print("{0:.2f}% 3 Ребенка".format(b4*100))
print("{0:.2f}% Многодетные".format(b5*100))

8.16% Нет детей
10.09% 1 Ребенок
10.49% 2 Ребенка
8.91% 3 Ребенка
10.81% Многодетные


Дети - всегда ответственность, везде и во всем. Так, у тех людей где есть хотя бы 1 ребенок шансы выплатить кредит больше, чем у бездетных (8,16 - процент должников)

In [34]:
#теперь про семейное положение посмотрим
banki_of_family_status = banki.groupby(['family_status','debt']).agg({'debt':["count"]})
banki_of_family_status

debt
                            count
family_status         debt       
Не женат / не замужем 0      2536
                      1       274
в разводе             0      1110
                      1        85
вдовец / вдова        0       896
                      1        63
гражданский брак      0      3762
                      1       388
женат / замужем       0     11408
                      1       931

In [35]:
data_pivot2 = banki.pivot_table(index=['debt'], columns='family_status', values='gender', aggfunc='count')
free = data_pivot2['Не женат / не замужем'][1] / data_pivot2['Не женат / не замужем'][0]
divorce = data_pivot2['в разводе'][1] / data_pivot2['в разводе'][0]
widow = data_pivot2['вдовец / вдова'][1] / data_pivot2['вдовец / вдова'][0]
partner = data_pivot2['гражданский брак'][1] / data_pivot2['гражданский брак'][0]
marriage = data_pivot2['женат / замужем'][1] / data_pivot2['женат / замужем'][0]
print("{0:.2f}% Не женат / не замужем".format(free*100))
print("{0:.2f}% в разводе".format(divorce*100))
print("{0:.2f}% вдовец / вдова".format(widow*100))
print("{0:.2f}% гражданский брак".format(partner*100))
print("{0:.2f}% женат / замужем".format(marriage*100))

10.80% Не женат / не замужем
7.66% в разводе
7.03% вдовец / вдова
10.31% гражданский брак
8.16% женат / замужем


Не женатые реже выплачивают кредиты в срок как и в гражданском браке. Самые низкие шансы стать проблемным заемщиком у вдовцов - всего 7,03, а самые высокие у не женатых и живущих в гражданском браке - 10.80% и 10.31% соответственно.

<div class="alert alert-block alert-success">
<b>✔️ Success:</b> все верно, но добавляй в выводы больше конкретики :) заказчику важны цифры

In [36]:
#теперь про зависимость от дохода
banki_income = banki.groupby(['total_income_category','debt']).agg({'debt':["count"]})
banki_income

debt
                            count
total_income_category debt       
A                     0        23
                      1         2
B                     0      4683
                      1       356
C                     0     14656
                      1      1360
D                     0       329
                      1        21
E                     0        20
                      1         2

In [37]:
data_pivot0 = banki.pivot_table(index=['debt'], columns='total_income_category', values='gender', aggfunc='count')
data_pivot0

in12 = data_pivot0['A'][1] / data_pivot0['A'][0]
in22 = data_pivot0['B'][1] / data_pivot0['B'][0]
in32 = data_pivot0['C'][1] / data_pivot0['C'][0]
in42 = data_pivot0['D'][1] / data_pivot0['D'][0]
in52 = data_pivot0['E'][1] / data_pivot0['E'][0]

print("{0:.2f}% A".format(in12*100))
print("{0:.2f}% B".format(in22*100))
print("{0:.2f}% C".format(in32*100))
print("{0:.2f}% D".format(in42*100))
print("{0:.2f}% E".format(in52*100))

8.70% A
7.60% B
9.28% C
6.38% D
10.00% E


In [38]:
# хуже всего креды выплачивают богачи

In [39]:
# Как разные цели кредита влияют на его возврат в срок?

data_pivot3 = banki.pivot_table(index=['debt'], columns='purpose_category', values='gender', aggfunc='count')
data_pivot3

in1 = data_pivot3['операции с автомобилем'][1] / data_pivot3['операции с автомобилем'][0]
in3 = data_pivot3['операции с недвижимостью'][1] / data_pivot3['операции с недвижимостью'][0]
in4 = data_pivot3['получение образование'][1] / data_pivot3['получение образование'][0]
in5 = data_pivot3['проведение свадьбы'][1] / data_pivot3['проведение свадьбы'][0]

print("{0:.2f}% операции с автомобилем".format(in1*100))
print("{0:.2f}% операции с недвижимостью".format(in3*100))
print("{0:.2f}% получение образование".format(in4*100))
print("{0:.2f}% проведение свадьбы".format(in5*100))

10.33% операции с автомобилем
7.80% операции с недвижимостью
10.16% получение образование
8.70% проведение свадьбы


In [40]:
# Лучше всего выплачивают кредиты если берут недвижку или на свадьбу

In [41]:
# Есть ли связь с образованием?
banki_education = banki.groupby(['education','debt']).agg({'debt':["count"]})
banki_education

debt
                          count
education           debt       
высшее              0      4972
                    1       278
начальное           0       251
                    1        31
неоконченное высшее 0       675
                    1        68
среднее             0     13808
                    1      1364
ученая степень      0         6

In [42]:
# как и следовало ожидать, обрахование влияет очень сильно, как на выплату кредита, так и по жизни )

### Ответы на вопросы.

Итак:

    Чаще всего целью кредита являются: недвижимость, автомобиль и реже образование;
    Чаще всего не выплачиваются кредиты, взятые автомобиль (10.33%) и образование (10.16%);
    Заёмщики с детьми, в целом чаще других выплачивали кредит вовремя против бездетных (8.16%);
    Заёмщики, которые были не женаты/замужем либо в гражданском браке чаще других не выплачивали кредит вовремя (10.80% и 10.31% соответственно);
    В целом, уровень образования влияет на выплату кредита в срок, чем выше уровень образования, тем больше шансы выплаты;
    Клиенты, с высоким доходом чаще других не выплачивали кредит вовремя.
    

На риск невыплаты кредита сильно влиет почти каждый фактор, это и количество детей, и доход, и семейное положение и цель кредита.

Однако, если у заёмщика доход высокий и он не женат, это еще не значит, что он не сможет выплатить вовремя взятый кредит. Нужно учитывать множетсво факторов, связанных с личностью клиента (может у него ученая степень и он берет на свадьбу).


<div class="alert alert-block alert-success">
<b>✔️ Success:</b> Отличная работа! Стоит немного "усилить" вывод цифрами и будет шикарно :)